### ISSUE TITLE
CI: Test that uninstall really uninstalls everything

### ISSUE THREAD

**Issue by RalfJung** [coq/opam-coq-archive#247]
Our CI could test that `opam unimstall` really uninstalls everything that got installed (and nothing more).

**Comment by mattam82** [comment]
Is opam supporting that somehow?

**Comment by clarus** [comment]
The opam bench tests that everything is uninstalled and nothing more. For example: https://coq-bench.github.io/clean/Linux-x86_64-4.05.0-2.0.1/released/8.5.0~camlp4/aac-tactics/8.5.1.html (bottom of the page, no missing or wrong removes). Currently, there are exactly zero uninstall errors! This kind of bug was largely solved by checking uninstalls in the bench, and 100% solved by opam 2 which does the uninstall itself.

There is still a frightening warning when uninstalling packages with an explicit `remove` directive and no `light-uninstall` flag:
```
~€ opam list
opam # Packages matching: installed
# Name              # Installed       # Synopsis
base-bigarray       base
base-num            base              Num library distributed with the OCaml compiler
base-threads        base
base-unix           base
camlp5              7.06.10-g84ce6cc4 Preprocessor-pretty-printer of OCaml
conf-m4             1                 Virtual package relying on m4
coq                 8.6               pinned to version 8.6
num                 0                 The Num library for arbitrary-precision integer and rational arithmetic
ocaml               4.05.0            The OCaml compiler (virtual package)
ocaml-base-compiler 4.05.0            Official 4.05.0 release
ocaml-config        1                 OCaml Switch Configuration
ocamlfind           1.8.0             A library manager for OCaml
~€ opam install coq-bellantonicook
The following actions will be performed:
  ∗ install coq-bellantonicook 1.0.0

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[coq-bellantonicook.1.0.0] found in cache

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
∗ installed coq-bellantonicook.1.0.0
Done.
~€ opam remove coq-bellantonicook
The following actions will be performed:
  ⊘ remove coq-bellantonicook 1.0.0

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[coq-bellantonicook.1.0.0] found in cache

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[WARNING] package uninstall script failed at /home/guillaume/opam.8.6/opam-init/hooks/sandbox.sh remove make uninstall:
          # context     2.0.4 | linux/x86_64 | ocaml-base-compiler.4.05.0 | file:///home/guillaume/Bureau/Opam/opam-coq-archive-clarus/released
          # path        ~/opam.8.6/4.05.0/.opam-switch/remove/coq-bellantonicook.1.0.0
          # command     ~/opam.8.6/opam-init/hooks/sandbox.sh remove make uninstall
          # exit-code   2
          # env-file    ~/opam.8.6/log/coq-bellantonicook-18010-05296d.env
          # output-file ~/opam.8.6/log/coq-bellantonicook-18010-05296d.out
          ### output ###
          # [...]
          # echo '#!/bin/sh' > uninstall_me.sh
          # printf 'cd "${DSTROOT}""/home/guillaume/opam.8.6/4.05.0/lib/coq/user-contrib"/BellantoniCook && rm -f    BC.vo BCBinary.vo BCI.vo BCI_to_BC.vo BCLib.vo BCUnary.vo BC_to_BCI.vo
          BC_to_Cobham.vo Bitstring.vo Cobham.vo CobhamBinary.vo CobhamLib.vo CobhamUnary.vo Cobham_to_BC.vo Lib.vo MultiPoly.vo && find . -type d -and -empty -delete\ncd
          "${DSTROOT}""/home/guillaume/opam.8.6/4.05.0/lib/coq/user-[...]
          # printf 'cd "${DSTROOT}""/home/guillaume/opam.8.6/4.05.0/share/doc/coq/user-contrib"/BellantoniCook \\\n' >> "uninstall_me.sh"
          # printf '&& rm -f \n' >> "uninstall_me.sh"
          # printf 'cd "${DSTROOT}""/home/guillaume/opam.8.6/4.05.0/share/doc/coq/user-contrib" && find BellantoniCook/html -maxdepth 0 -and -empty -exec rmdir -p \{\} \;\n' >>
          "uninstall_me.sh"
          # chmod +x uninstall_me.sh
          # sh uninstall_me.sh
          # uninstall_me.sh: 4: cd: can't cd to /home/guillaume/opam.8.6/4.05.0/share/doc/coq/user-contrib/BellantoniCook
          # uninstall_me.sh: 6: cd: can't cd to /home/guillaume/opam.8.6/4.05.0/share/doc/coq/user-contrib
          # make[1]: *** [uninstall] Erreur 2
          # make[1]: quittant le répertoire « /home/guillaume/opam.8.6/4.05.0/.opam-switch/remove/coq-bellantonicook.1.0.0 »
          # make: *** [uninstall] Erreur 2

Done.
~€ echo $?
0
```
For the user it seems like it failed, but it actually did not. We should probably solve that.

**Comment by clarus** [comment]
I am closing this issue as it is old and opam 2 now largely automates the uninstall, preventing most errors.
